### PR TITLE
 Added Widevine Base License Acquisition + FairPlay support

### PR DESCRIPTION
**Pull request by rnrneverdies**
We also included a new sample for FairPlay.

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@rnrneverdies__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <span>We will now validate the agreement and then real humans will evaluate your PR.</span>

TTYL, AZPRBOT;

**Comment by yaqiyang** [comment]
Thanks for the contribution. I will take a look in the next few days.

File: .gitignore [hunk 1]
```diff
@@ -9,3 +9,4 @@ vendor
 nbproject
 phpcs-checkstyle.tmp
 composer.phar
+.vscode/
```
Comment (phansys) [inline .gitignore]: Why is this change required? Environment specific files/directories should be added to your local ignore list: https://help.github.com/articles/ignoring-files/#create-a-local-gitignore

File: examples/MediaServices/vodworkflow_aes.php [hunk 2]
```diff
@@ -331,7 +331,7 @@ function publishEncodedAsset($restProxy, $encodedAsset)
     $locator = $restProxy->createLocator($locator);
 
     // 6.4 Create a Smooth Streaming base URL
-    $stremingUrl = $locator->getPath().$manifestFile->getName().'/manifest';
+    $stremingUrl = $locator->getPath().$manifestFile->getName()."/manifest";
```
Comment (phansys) [inline examples/MediaServices/vodworkflow_aes.php]: How is this change related to Widevine / FairPlay?

File: src/MediaServices/Models/AssetDeliveryPolicyConfigurationKey.php [hunk 3]
```diff
@@ -96,6 +96,35 @@ class AssetDeliveryPolicyConfigurationKey
      */
     const WIDEVINE_LICENSE_ACQUISITION_URL = 7;
 
+	/**
```
Comment (phansys) [inline src/MediaServices/Models/AssetDeliveryPolicyConfigurationKey.php]: Please, fix indentation.

File: src/MediaServices/Templates/FairPlayConfiguration.php [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    public $ContentEncryptionIV;
+
+    public static function createSerializedFairPlayOptionConfiguration($cert, $pkey, $pfxPassword, $pfxPasswordKeyId, $askId, $contentIv) {
+
+        openssl_pkcs12_export($cert, $certBytes, $pkey, $pfxPassword);
+        $certString = base64_encode($certBytes);
+        $template = new FairPlayConfiguration();
+        $template->ASkId = $askId;
+        $template->ContentEncryptionIV = $contentIv;
+		$template->FairPlayPfx = $certString;
```
Comment (phansys) [inline src/MediaServices/Templates/FairPlayConfiguration.php]: Please, fix indentation.

File: src/MediaServices/Templates/FairPlayConfiguration.php [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+		$template->FairPlayPfxPasswordId = $pfxPasswordKeyId;
+    
+        return json_encode($template);
+    }
+   
+}
+
```
Comment (phansys) [inline src/MediaServices/Templates/FairPlayConfiguration.php]: Please, remove surplus line breaks at EOF.

File: examples/MediaServices/vodworkflow_aes.php (hunk 2 above)
Comment (rnrneverdies) [inline examples/MediaServices/vodworkflow_aes.php]: It's not, but we went forced to applying many patches because someone has refactored the entire repository, this one seems that is stowaway from the older branch.

Comment (rnrneverdies) [inline examples/MediaServices/vodworkflow_aes.php]: fixed

File: src/MediaServices/Models/AssetDeliveryPolicyConfigurationKey.php (hunk 3 above)
Comment (rnrneverdies) [inline src/MediaServices/Models/AssetDeliveryPolicyConfigurationKey.php]: fixed

File: src/MediaServices/Templates/FairPlayConfiguration.php (hunk 4 above)
Comment (rnrneverdies) [inline src/MediaServices/Templates/FairPlayConfiguration.php]: fixed

File: src/MediaServices/Templates/FairPlayConfiguration.php (hunk 5 above)
Comment (rnrneverdies) [inline src/MediaServices/Templates/FairPlayConfiguration.php]: fixed

**Comment by yaqiyang** [comment]
I got two problems when I ran your code against the unit tests.
First, I am getting this error.
  <error type="PHPUnit_Framework_Error_Warning">Tests\unit\WindowsAzure\MediaServices\MediaServicesRestProxyTest::testCreateContentKey
mcrypt_encrypt(): Size of key is too large for this algorithm

D:\Repos\azure-sdk-for-php\src\MediaServices\Models\ContentKey.php:427
D:\Repos\azure-sdk-for-php\src\MediaServices\Models\ContentKey.php:453
D:\Repos\azure-sdk-for-php\tests\unit\WindowsAzure\MediaServices\MediaServicesRestProxyTest.php:1294
</error>
It means mcrypt_encrypt() is failing with your change at src\MediaServices\Models\ContentKey.php:427. Does this change require configuration changes?

Second, you changed src/MediaServices/Templates/TokenRestrictionTemplateSerializer.php which also cause a test to fail (Tests\unit\WindowsAzure\MediaServices\Templates\TokenRestrictionTemplateSerializerTest::testSerializerSWT) because the serialized string is now different. Will this change cause any regressions for existing users?

Thanks
Yaqi Yang

**Comment by rnrneverdies** [comment]
Hi @yaqiyang, sorry for the delay.

I'm going to take a look on your comments asap and will let you know. 

Thanks.

**Comment by rnrneverdies** [comment]
Hi @yaqiyang, I've updated this branch with the following changes:

 - Reverted the changes on the token restriction template to keep it backwards compatible.
 - Updated content key checksum generation algorithm to make it optional and to keep it backwards compatible.

Please let me know if you have additional feedback.

Thanks.

**Comment by yaqiyang** [comment]
Thanks. I will take a look once I come back from vacation next week.
